Making ops c10-full: Storage arguments

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49259 Remove `use_c10_dispatcher: full` lines
* #49251 Remove generated_unboxing_wrappers and setManuallyBoxedKernel
* #49220 Remove .impl_UNBOXED() and functionalities associated with it
* #49164 Remove codegen logic to support non-c10-full ops
* #49138 [wip] Making ops c10-full: list of optional tensors
* #49463 Remove set_quantizer_ from native_functions.yaml
* **#49146 Making ops c10-full: Storage arguments**

Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

smessmer committed:
Making ops c10-full: Storage arguments

Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

ezyang reviewed (approved):
It's probably worth noting that TorchScript frontend probably doesn't actually support storage haha

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Storage arguments"


Add support for Storage arguments to IValue and the JIT typing system, and make ops that were blocked on that c10-full.

Differential Revision: [D25456799](https://our.internmc.facebook.com/intern/diff/D25456799/)

[ghstack-poisoned]

swolchok commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -314,6 +315,20 @@ struct CAFFE2_API IValue final {
     return static_cast<at::TensorImpl*>(payload.as_intrusive_ptr);
   }
 
+  IValue(at::Storage s) : tag(Tag::Storage), is_intrusive_ptr(static_cast<bool>(s)) {
```
Do we really need to conditionally set `is_intrusive_ptr` here? It looks to me like empty `Storage` doesn't have a specialized singleton representation; it's just `nullptr`
